### PR TITLE
Update Target Based Resolution

### PR DESCRIPTION
**Pull request by SDGGiesbrecht**
Associated pull requests:

- [x] #3858
- [x] #3860
- [x] #3862
- [x] #3871
- [x] #3870
- [x] #3877
- [ ] #4038
- [x] Solve [crash in `testTransitiveDependencySwitchWithSameIdentity`](https://github.com/apple/swift-package-manager/pull/4038#issuecomment-1033388208)
- [ ] Fix remaining test expectation differences.
- [ ] Verify that there is no remaining diff
- [ ] Lastly, activate `ENABLE_TARGET_BASED_DEPENDENCY_RESOLUTION` by default (#3880)

***

See [this comment](https://github.com/apple/swift-package-manager/pull/3829#issuecomment-958517169) in #3829 for the impetus.

I have brought all but two tests up to date, and now I have questions about them. (For one of @tomerd, @abertelrud or @neonichu?)

1. `FunctionalTests.PluginTests/testUseOfBuildToolPluginProductByExecutableAcrossPackages` is trying to launch a plugin executable _before_ building it. Can someone point me roughly to where or how SwiftPM communicates to llbuild that one command cannot start until another has finished? (All the necessary commands seem to be there, just not not in the right order.)
2. `XCBuildSupportTests.PIFBuilderTests/testTestProducts` is objecting with “failed to retrieve search paths with pkg-config; maybe pkg-config is not installed”. I do not know whether that is a real failure or just due to the machine I am using. Can someone else confirm one way or the other?

(I will list and explain all the other updates once I am finished; all are minor and self‐contained.)

### CODE REVIEW

**Comment by abertelrud** [comment]
> FunctionalTests.PluginTests/testUseOfBuildToolPluginProductByExecutableAcrossPackages is trying to launch a plugin executable before building it. Can someone point me roughly to where or how SwiftPM communicates to llbuild that one command cannot start until another has finished? (All the necessary commands seem to be there, just not not in the right order.)

Looks as if this plugin depends on an executable built from another target, and that is the one that's failing to build first?  The plugin dependencies use the normal dependencies machinery underneath, so in theory this should already provide the right dependencies.  I'll take a closer looks and see if I have some advice.

> XCBuildSupportTests.PIFBuilderTests/testTestProducts is objecting with “failed to retrieve search paths with pkg-config; maybe pkg-config is not installed”. I do not know whether that is a real failure or just due to the machine I am using. Can someone else confirm one way or the other?

I thought that `pkg-config` was handled through a custom parser rather than invoking the actual `pkg-config` tool, so this puzzles me.  I can't confirm but will take a look at this as well.

**Comment by SDGGiesbrecht** [comment]
Thanks, @abertelrud.

I have already been debugging the plugin one some more.

I have discovered that `computeDependencies(of: ResolvedProduct)` seems never to be called for any plugin _product_, even when it is in the root package. [This](https://github.com/apple/swift-package-manager/blob/52c852bec25c4b049c4ed4657eb36294dc75daff/Sources/Build/BuildPlan.swift#L1611) looks like the line that filters them out beforehand. I suspect in‐package plugins work okay, because even the targets and their interdependencies are intact, even though no product is ever “built”. It might be correct to “build nothing”, but culling the node there is also culling its transitive dependencies. At the moment I am experimenting with deferring the culling to a later stage, but I keep getting other errors because at present, later code does not expect plugins and triggers unrelated fatal errors. I will get back to you when I have investigated more thoroughly.

**Comment by SDGGiesbrecht** [comment]
That was probably a wild goose chase. Adding the plug‐in tool product as a dependency in addition to the plug‐in script restores the missing links in the chain and llbuild knows what to do. But it does not look like that extra declaration is supposed to be necessary. Back to square one.

**Comment by SDGGiesbrecht** [comment]
I figured out the problem with the plug‐ins. The part that creates the executable comes from the _product_, not the _target_. Since the declared dependencies go `client` → `plugin‐script‐product` → `plugin‐script‐target` → `executable‐target`, there is no `executable‐product` in the graph. I fixed it by synthesizing a product for any executable that a plug‐in script depends on (funnelling it into the same code that does the synthesis for executable targets in the root package). I am doing further tests now, and will commit it once I know if it has any unexpected side‐effects.

(After figuring that out, I went back to `main`, removed the explicit product from the test fixture, and ran the test. It failed in the same way with the same cryptic message. So it will need fixing even if you decide not to keep any of this target‐based stuff.)

**Comment by SDGGiesbrecht** [comment]
In addition to the `pkd-config` puzzle, I have noticed some `FIXME`s I would like to look closer at, because they claim to be related to target based resolution. I will probably get to it next week.

**Comment by SDGGiesbrecht** [comment]
I went searching in the wild until I found a broken package and then traced it through the resolver. `PubGrubPackageContainer`’s private cache was indexing incorrectly, and it was capable of derailing things downstream in certain situations. My latest commit fixed it and added a test.

Reading back over the earlier threads, this bug was probably the cause of the issues @neonichu and I were trying to solve months ago. The symptoms certainly match.

At this point, it would be helpful to know how this fares against smoke testing and the source compatibility suite.

**Comment by abertelrud** [comment]
> I figured out the problem with the plug‐ins. The part that creates the executable binary comes from the _product_, not the _target_. Since the declared dependencies go `client` → `plugin‐script‐product` → `plugin‐script‐target` → `executable‐target`, there is no `executable‐product` in the graph. I fixed it by synthesizing a product for any executable that a plug‐in script depends on (funnelling it into the same code that does the synthesis for executable targets in the root package). I am doing further tests now, and will commit it once I know whether it has any unexpected side‐effects.
> 
> (After figuring that out, I went back to `main`, removed the explicit product from the test fixture, and ran the test. It failed in the same way with the same cryptic message. So it will need fixing even if you decide not to keep any of this target‐based stuff.)

Wow, thank you for digging into this and finding this!  I'll take a look at the specifics of what you had to do with synthesizing a product, but it sounds like the right approach — I actually thought that all packages already had this synthesis.  The only thing to watch out for is that dependencies shouldn't have access to the synthesized executables if the package author didn't intentionally make them public.

The product vs target distinction did cause a certain amount of grief when the plugin support was added (as it also did when testable executable modules were addded).

(By the way, the product vs target distinction may have seemed like a good idea initially but at this point costs everyone more than it's worth, and I think is worth revisiting at some point — but that is obviously a later topic for a discussion and proposal.)

**Comment by abertelrud** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -315,7 +315,9 @@ private func createResolvedPackages(
             }
         }
 
-        packageBuilder.dependencies = Array(dependencies.values)
+        packageBuilder.dependencies = dependencies.values
+            // Restore determinism:
+            .sorted(by: { $0.package.identity < $1.package.identity })
```
Comment (abertelrud) [inline Sources/PackageGraph/PackageGraph+Loading.swift:320]: Sorting these for stability is great, I think.  In the long run I think it would actually be nice if we could get back to using the order in which dependencies were declared in the manifest.  It shouldn't make a difference to the logic of course, but it seems to me that using the user's given order would be more natural to read as a package author.  It's a total detail, just something I hope that SwiftPM can get back to at some point (I think it used to do this early on).

File: Sources/PackageLoading/PackageBuilder.swift [hunk 2]
```diff
@@ -1253,37 +1253,44 @@ public final class PackageBuilder {
...
-                    continue
-                } else {
+        let productMap = products.reduce(into: [String: Product]()) { partial, iterator in
+            partial[iterator.key] = iterator.item
+        }
+
+        let implicitPlugInExecutables = Set(targets.lazy
+             .filter({ $0.type == .plugin })
+             .flatMap({ $0.dependencies })
+             .map { $0.name })
```
Comment (abertelrud) [inline Sources/PackageLoading/PackageBuilder.swift:1277]: If I read this correctly, this prevents non-plugin targets from depending on these implicit products (which would be great).  Or am I reading that wrong?

Comment (SDGGiesbrecht) [reply Sources/PackageLoading/PackageBuilder.swift:1277]: Probably...? It certainly means they are not there to depend on during resolution or loading. But later when the build system is constructed there is no insulation to prevent dangling references from pointing where it will end up. What I do not know for sure, is if there is a way for it to sneak under the diagnostic checks to get that far, since some dangling references are temporarily permitted so that later diagnostics have more information available to suggest fixes.

File: Sources/PackageGraph/PackageGraph+Loading.swift (hunk 1 above)
Comment (SDGGiesbrecht) [reply Sources/PackageGraph/PackageGraph+Loading.swift:320]: I thought about that too, and considered refactoring to use an `OrderedDictionary`. But pulling in `swift-collections` seemed too far off‐topic for this PR. It is certainly something to consider in the future.

**Comment by SDGGiesbrecht** [comment]
@abertelrud,

> The only thing to watch out for is that dependencies shouldn't have access to the synthesized executables if the package author didn't intentionally make them public.

To make sure I understand what you mean, let me ask it like this: Should the following work or not, assuming the intent is to have SwiftPM to build `PlugInExecutableTarget` and launch it like `PlugInExecutableTarget some-argument --some-flag`?

```swift
// Dependency
let package = Package(
  // ...
  products: [
    .plugin(name: "PlugInScriptProduct", targets: ["PlugInScriptTarget"]),
    // (There is no product for “PlugInExecutableTarget”.)
  ],
  targets: [
    .plugin(
      name: "PluginScriptTarget",
      // ...
      dependencies: ["PlugInExecutableTarget"]
    ),
    .executableTarget(name: "PlugInExecutableTarget")
  ]
)
```

```swift
// Root
let package = Package(
  // ...
  targets: [
    .target(
      // ...
      plugins: [
        .plugin(name: "PlugInScriptProduct", package: "Dependency")
      ]
    )
  ]
)
```

**Comment by SDGGiesbrecht** [comment]
Well the CI log tells me _half_ of the pkg-config issue is real, since CI did not wine about it the installation the way it does on my machine, but it did hit other trouble later. I will keep investigating.

**Comment by SDGGiesbrecht** [comment]
I figured out the `pkg‐config` issue. The in-memory file system just lacked the `.pc` file. The test probably just fell out of date while it was disabled.

Looking over other `FIXME` now.

**Comment by SDGGiesbrecht** [comment]
That last commit fixes a bug where packages with tools version from 5.2 to 5.4 that had a name property different from the last path component behaved like 5.1 manifests do, asking the resolver to look for more than necessary. This was because the identity work broke the lookup from the target dependency to the package dependency, and the manifest could not tell which package the product was supposed to belong to. Due to having no known connection, the logic was landing in the code branches intended for resolving 5.1 manifests.

Locally all tests are green and every package have tried now seems to be working as expected. Let me know:
- if you find any other problems
- if you want this PR rebased, split up, or otherwise adjusted
- what you want done with `ENABLE_TARGET_BASED_DEPENDENCY_RESOLUTION`:
  - remove the checks entirely?
  - default to on like the PR’s current state (so the central switch remains for a while in case problems arise)?
  - default to off like it has been for the last while?
  - or some other plan?

**Comment by neonichu** [comment]
Thanks for your work, @SDGGiesbrecht!

I think leaving `ENABLE_TARGET_BASED_DEPENDENCY_RESOLUTION` in for a bit, but turned on sounds like the best option to me for this PR. Once we're confident, we can remove it entirely in a future PR.

**Comment by SDGGiesbrecht** [comment]
I adjusted the manifest so that you can set `DISABLE_TARGET_BASED_DEPENDENCY_RESOLUTION` in the environment to flip the compilation condition on or off. By default, the feature is enabled.

File: Sources/PackageGraph/PackageGraph+Loading.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackageGraph/PackageGraph+Loading.swift:320]: Absolutely.  It was more of a comment about a general approach here.  I should have made that clear.

File: Sources/PackageLoading/PackageBuilder.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/PackageLoading/PackageBuilder.swift:1277]: Thanks for clarifying — it's very possible that we should (in a separate PR, by me) add more checks for that kind of thing.

Representing plugins as targets definitely has some trade-offs.  It's a bit odd to do so, but making them completely separate would have made it more difficult to fit them into the dependency graph.  I still think it would be even more complicated to make them participate in the dependency graph without being targets, but it definitely does introduce some edge cases.

**Comment by abertelrud** [comment]
> @abertelrud,
> 
> > The only thing to watch out for is that dependencies shouldn't have access to the synthesized executables if the package author didn't intentionally make them public.
> 
> To make sure I understand what you mean, let me ask it like this: Should the following work or not, assuming the intent is to have SwiftPM to build `PlugInExecutableTarget` and launch it like `PlugInExecutableTarget some-argument --some-flag`?
> 
> ```swift
> // Dependency
> let package = Package(
>   // ...
>   products: [
>     .plugin(name: "PlugInScriptProduct", targets: ["PlugInScriptTarget"]),
>     // (There is no product for “PlugInExecutableTarget”.)
>   ],
>   targets: [
>     .plugin(
>       name: "PluginScriptTarget",
>       // ...
>       dependencies: ["PlugInExecutableTarget"]
>     ),
>     .executableTarget(name: "PlugInExecutableTarget")
>   ]
> )
> ```
> 
>  ```swift
> // Root
> let package = Package(
>   // ...
>   targets: [
>     .target(
>       // ...
>       plugins: [
>         .plugin(name: "PlugInScriptProduct", package: "Dependency")
>       ]
>     )
>   ]
> )
> ```

Thanks for laying this out in such detail.  Yes, this should work, since it's the plugin inside the package that depends on an executable inside the package.  What I think should not happen would be for `PlugInExecutableTarget` to be accessible from the root package (by virtue of being reachable through a plugin).

**Comment by abertelrud** [comment]
> Thanks for your work, @SDGGiesbrecht!

I agree, thank you very much @SDGGiesbrecht!

> I think leaving `ENABLE_TARGET_BASED_DEPENDENCY_RESOLUTION` in for a bit, but turned on sounds like the best option to me for this PR. Once we're confident, we can remove it entirely in a future PR.

This seems reasonable to me we as well.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by SDGGiesbrecht** [comment]
> Thanks for laying this out in such detail. Yes, this should work, since it's the plugin inside the package that depends on an executable inside the package. What I think should not happen would be for PlugInExecutableTarget to be accessible from the root package (by virtue of being reachable through a plugin).

Then we were thinking the same thing. The product does not exist yet during resolution, so declaring a direct dependency on it from another package should fail. The product is only created later during loading so that the build system can use it.

**Comment by abertelrud** [comment]
> > Thanks for laying this out in such detail. Yes, this should work, since it's the plugin inside the package that depends on an executable inside the package. What I think should not happen would be for PlugInExecutableTarget to be accessible from the root package (by virtue of being reachable through a plugin).
> 
> Then we were thinking the same thing. The product does not exist yet during resolution, so declaring a direct dependency on it from another package should fail. The product is only created later during loading so that the build system can use it.

That sounds perfect.  Thanks!

**Comment by abertelrud** [comment]
> I figured out the problem with the plug‐ins. The part that creates the executable binary comes from the _product_, not the _target_. Since the declared dependencies go `client` → `plugin‐script‐product` → `plugin‐script‐target` → `executable‐target`, there is no `executable‐product` in the graph. I fixed it by synthesizing a product for any executable that a plug‐in script depends on (funnelling it into the same code that does the synthesis for executable targets in the root package). I am doing further tests now, and will commit it once I know whether it has any unexpected side‐effects.
> 
> (After figuring that out, I went back to `main`, removed the explicit product from the test fixture, and ran the test. It failed in the same way with the same cryptic message. So it will need fixing even if you decide not to keep any of this target‐based stuff.)

Thanks again for all your work here @SDGGiesbrecht.  Would it be possible to split out the plugin fix, or point me to what needs to be factored out, so that can be tested and landed independently, and then we figure out how and when to land the rest of your changes separately?

**Comment by SDGGiesbrecht** [comment]
Sure. In a separate PR I will create a test that is independent of the resolution mode and then steal from here as much as ends up being necessary.

**Comment by abertelrud** [comment]
> Sure. In a separate PR I will create a test that is independent of the resolution mode and then steal from here as much as ends up being necessary.

Thank you so much for doing that!  Very much appreciated.  Splitting them up should also make it easier to diagnose any potential fallout later.

**Comment by SDGGiesbrecht** [comment]
I think there is more that I can split out bug by bug to have tested and looked at irrespective of the resolution mode. I will keep this PR branch up to date as a proof of concept and a gauge of the remaining diff, but I feel like merging the pieces incrementally would make it easier to walk back only the relevant piece if something unexpected is discovered later.

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by SDGGiesbrecht** [comment]
Hey, does that mean someone granted me access to...

**Comment by SDGGiesbrecht** [comment]
@swift-ci please smoke test

**Comment by SDGGiesbrecht** [comment]
I guess not. Oh, well.

**Comment by SDGGiesbrecht** [comment]
I am trying to update this to resolve the merge conflicts, and there is a new issue in `testResolutionMixedRegistryAndSourceControl2`.

The test has:
- A package dependency with `.registry(identity: "org.foo", [...])`
- A target dependency with `.product([...], package: "foo")`

@tomerd, Is that supposed to be a valid reference or not? `packageDependencies(referencedBy:)` currently rejects it because its `nameForTargetDependencyResolutionOnly` is `org.foo` instead.

- If the reference is supposed to be valid, then I need to fix `nameForTargetDependencyResolutionOnly`.
- If the reference is supposed to be invalid, then all I need to do is add a valid reference to the test so that it does not get culled and the diagnostics the test cares about are not skipped.

**Comment by tomerd** [comment]
thanks for looking into this @SDGGiesbrecht, these set of tests are a bit tricky as they test edge cases of duplicate sim / registry dependences. 

this specific test is even more edge-casey as it tests duplicates at the root level,

the root package has:

```
MockPackage(
                    name: "Root",
                    path: "root",
                    targets: [
                        MockTarget(name: "RootTarget", dependencies: [
                            .product(name: "FooProduct", package: "foo")
                        ]),
                    ],
                    products: [],
                    dependencies: [
                        .sourceControl(url: "https://git/org/foo", requirement: .upToNextMajor(from: "1.0.0")),
                        .registry(identity: "org.foo", requirement: .upToNextMajor(from: "1.0.0")),
                    ],
                    toolsVersion: .v5_6
                ),
```

which is [deliberate] duplicate mixed entries since `https://git/org/foo` ==  `org.foo` as defined in this test

so the target dependency `.product(name: "FooProduct", package: "foo")` should be valid since it is pointing to `https://git/org/foo` which is in the graph.

what error do you see?

**Comment by SDGGiesbrecht** [comment]
> what error do you see?

The expected errors vanish, and this warning appears instead:

> dependency 'org.foo' is not used by any target

The resolver does not think any target uses `org.foo`, because its `nameForTargetDependencyResolutionOnly` is `org.foo` which is not a case‐sensitive match for the `foo` in the reference. See the source [here](https://github.com/apple/swift-package-manager/blob/e3432e9801528a0433c2d0613e4f693144d03d44/Sources/PackageModel/Manifest.swift#L308-L315).

This might also be broken even in legacy resolution. Right now the test might only be succeeding because it happens to be in a root manifest, which bypasses lookup altogether (because in legacy it does not need to figure out its product filters). Transitive manifests still use `packageDependency(referencedBy:)`. See the source [here](https://github.com/apple/swift-package-manager/blob/e3432e9801528a0433c2d0613e4f693144d03d44/Sources/PackageModel/Manifest.swift#L190-L203).

**Comment by SDGGiesbrecht** [comment]
> valid unless they are de-duped, which is what happens when the test enables the de-duplication modes

The de‐duplication you speak of does not seem to be happening at a low enough level. Nothing in `Manifest`’s reference lookup is aware of these different modes.

But now that I know understand how it is intended to work, I think I will be able to find and sink the logic in order to fix it. Thank you.

**Comment by SDGGiesbrecht** [comment]
Never mind, I see now that the manifest is being completely overwritten and thus does not need to know about any of the higher level stuff.

I also neglected to notice earlier that the `.disabled` mode gets tested first, so all the print statements and break points I was looking at were under the opposite mode. 🤦 It was a red herring.

I am now pretty sure everything is working properly, and the test failure was just because the message changed along with the particular downstream error. Both the old and new errors indicate that it successfully navigated the stuff the test is about before any problems occurred.

**Comment by SDGGiesbrecht** [comment]
@elsh, would you double check [this last commit](https://github.com/apple/swift-package-manager/pull/3838/commits/1f27f39152f36b14c63dc2ebeb2ffda2c3276d00) for me? These two tests started failing after #5635. Part of the tests’ dependency graphs is unreachable, and thus only present under legacy resolution. Prior to your recent update, I had [simply added the missing links necessary to ensure the entire graph was reachable](https://github.com/apple/swift-package-manager/pull/3838/commits/0f092d748b9f5737037e52fdde60623bc7753535). But since the update, that no longer works.<sup>†</sup> So my latest commit instead just adapts the expectations to what the current manifests are actually asking for. What I want to know from you is if the tests still test what they are supposed to, or if the tests become meaningless because the relevant part was in the unreachable portion of the dependency graph that is now skipped entirely.

**Comment by SDGGiesbrecht** [comment]
Closing. This and related PRs are stale and I do not have time to maintain them anymore. I will leave the related branches around for now in case anyone else wants to pick them up.
